### PR TITLE
Refactor to new framework

### PR DESCRIPTION
**Pull request by catalinnnn**

### CODE REVIEW

File: elrond_dex_pair/src/lib.rs [hunk 1]
```diff
@@ -505,122 +487,111 @@ pub trait Pair {
...
-            &token_in,
-            &token_out,
-        );
+        self.update_reserves(&reserve_token_in, &reserve_token_out, &token_in, &token_out);
+        self.send_tokens(&token_out, &amount_out, &caller);
+        self.send_tokens(&token_in, &residuum, &caller);
 
         //The transaction was made. We are left with $(fee) of $(token_in) as fee.
-        if self.fee().is_enabled() {
+        if self.is_enabled() {
```
Comment (claudiulataretu) [inline elrond_dex_pair/src/lib.rs:531]: Same as above regarding fee

File: elrond_dex_pair/src/lib.rs [hunk 2]
```diff
@@ -942,62 +898,60 @@ pub trait Pair {
     }
 
     #[view(getAmountIn)]
-    fn get_amount_in(
+    fn get_amount_in_view(
```
Comment (claudiulataretu) [inline elrond_dex_pair/src/lib.rs:901]: Maybe you ca use cameCase in here too

File: elrond_dex_pair/src/lib.rs [hunk 3]
```diff
@@ -1022,15 +976,19 @@ pub trait Pair {
         self.state().get() == State::Active
     }
 
+    #[view]
+    fn getLpTokenIdentifier(&self) -> TokenIdentifier {
```
Comment (claudiulataretu) [inline elrond_dex_pair/src/lib.rs:980]: You can put this view function directly on the storage mapper

File: elrond_dex_pair/src/lib.rs [hunk 4]
```diff
@@ -451,33 +438,28 @@ pub trait Pair {
...
-            &reserve_token_out,
-            &token_in,
-            &token_out,
-        );
+        self.update_reserves(&reserve_token_in, &reserve_token_out, &token_in, &token_out);
+        self.send_tokens(&token_out, &amount_out_optimal, &caller);
 
         //The transaction was made. We are left with $(fee) of $(token_in) as fee.
-        if self.fee().is_enabled() {
+        if self.is_enabled() {
```
[the remark this answers is not in the document]
Comment (catalinnnn) [reply elrond_dex_pair/src/lib.rs]: Fixed

File: elrond_dex_pair/src/lib.rs (hunk 1 above)
Comment (catalinnnn) [reply elrond_dex_pair/src/lib.rs]: Fixed

File: elrond_dex_pair/src/lib.rs [hunk 5]
```diff
@@ -831,106 +792,101 @@ pub trait Pair {
...
+        let second_token_id = self.second_token_id().get();
+        let first_token_reserve = self.pair_reserve(&first_token_id).get();
+        let second_token_reserve = self.pair_reserve(&second_token_id).get();
+        let total_supply = self.total_supply().get();
         (first_token_reserve, second_token_reserve, total_supply).into()
     }
 
     #[view(getAmountOut)]
-    fn get_amount_out(&self, token_in: TokenIdentifier, amount_in: BigUint) -> SCResult<BigUint> {
+    fn get_amount_out_view(
```
[the remark this answers is not in the document]
Comment (catalinnnn) [reply elrond_dex_pair/src/lib.rs]: Fixed

File: elrond_dex_pair/src/lib.rs (hunk 2 above)
Comment (catalinnnn) [reply elrond_dex_pair/src/lib.rs]: Done

File: elrond_dex_pair/src/lib.rs (hunk 3 above)
Comment (catalinnnn) [reply elrond_dex_pair/src/lib.rs:974]: Unfortunately, you cannot

File: elrond_dex_farm/abi/src/main.rs [hunk 6]
```diff
@@ -1,7 +1,3 @@
-use elrond_dex_farm::*;
-use elrond_wasm_debug::*;
-
 fn main() {
-	let contract = FarmImpl::new(TxContext::dummy());
-	print!("{}", abi_json::contract_abi(&contract));
+	elrond_wasm_debug::abi_json::print_abi::<elrond_dex_farm::AbiProvider>();
 }
```
Comment (sasurobert) [inline elrond_dex_farm/abi/src/main.rs:3]: add empty line

File: elrond_dex_pair/abi/src/main.rs [hunk 7]
```diff
@@ -1,7 +1,3 @@
-use elrond_dex_pair::*;
-use elrond_wasm_debug::*;
-
 fn main() {
-	let contract = PairImpl::new(TxContext::dummy());
-	print!("{}", abi_json::contract_abi(&contract));
+	elrond_wasm_debug::abi_json::print_abi::<elrond_dex_pair::AbiProvider>();
 }
```
Comment (sasurobert) [inline elrond_dex_pair/abi/src/main.rs:3]: add an empty line

File: elrond_dex_pair/src/amm.rs [hunk 8]
```diff
@@ -1,82 +1,87 @@
...
 
-    fn get_special_fee_from_optimal_input(&self, amount_in_optimal: BigUint) -> BigUint {
+    fn get_special_fee_from_optimal_input(
+        &self,
+        amount_in_optimal: &Self::BigUint,
+    ) -> Self::BigUint {
         let amount_in_zero_fee = amount_in_optimal
-            * BigUint::from(100000 - self.total_fee_precent().get())
-            / BigUint::from(100000u64);
+            * &Self::BigUint::from(100000 - self.total_fee_percent().get())
```
Comment (sasurobert) [inline elrond_dex_pair/src/amm.rs:72]: *& ? for the first bigUint ? could be deleted.

File: elrond_dex_pair/src/liquidity_pool.rs [hunk 9]
```diff
@@ -58,6 +46,10 @@ pub trait LiquidityPoolModule {
             &liquidity,
         );
 
+        let mut total_supply = self.total_supply().get();
+        total_supply += liquidity.clone();
+        self.total_supply().set(&total_supply);
```
Comment (sasurobert) [inline elrond_dex_pair/src/liquidity_pool.rs:51]: was this a bug before ?

File: elrond_dex_pair/src/liquidity_pool.rs [hunk 10]
```diff
@@ -91,53 +83,58 @@ pub trait LiquidityPoolModule {
...
         self.send().esdt_local_burn(
             self.blockchain().get_gas_left(),
             lp_token_identifier.as_esdt_identifier(),
             &liquidity,
         );
 
+        let mut total_supply = self.total_supply().get();
+        require!(total_supply > liquidity, "Not enough supply");
+        total_supply -= liquidity;
+        self.total_supply().set(&total_supply);
```
Comment (sasurobert) [inline elrond_dex_pair/src/liquidity_pool.rs:115]: was this a bug before ?

**Review by claudiulataretu — APPROVED** [review]

File: elrond_dex_pair/src/lib.rs [hunk 11]
```diff
@@ -166,28 +169,21 @@ pub trait Pair {
...
         let lp_token_id = self.lp_token_identifier().get();
-        let liquidity = sc_try!(self.liquidity_pool().mint(
+        let liquidity = self.mint(
             first_token_amount.clone(),
             second_token_amount.clone(),
             lp_token_id.clone(),
-        ));
+        )?;
 
         let caller = &self.blockchain().get_caller();
```
Comment (sasurobert) [inline elrond_dex_pair/src/lib.rs:186]: ok. so you moved the total_supply computation inside the burn / mint

File: elrond_dex_pair/src/amm.rs (hunk 8 above)
Comment (catalinnnn) [reply elrond_dex_pair/src/amm.rs:72]: amount_in_optimal is a reference so the second argument of the multiply function should also be a reference

File: elrond_dex_pair/src/liquidity_pool.rs (hunk 10 above)
Comment (catalinnnn) [reply elrond_dex_pair/src/liquidity_pool.rs:115]: no

**Review by dorin-iancu — APPROVED** [review]

**Pull request merged**